### PR TITLE
[libc][stdfix] Add round functions for fixed point types.

### PR DESCRIPTION
**Pull request by lntue**

### CODE REVIEW

File: libc/src/__support/fixed_point/CMakeLists.txt [hunk 1]
```diff
@@ -17,4 +17,5 @@ add_header_library(
     libc.include.llvm-libc-macros.stdfix_macros
     libc.src.__support.macros.attributes
     libc.src.__support.macros.optimization
+    libc.src.__support.CPP.bit
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/CMakeLists.txt:20]: You may want to update the Bazel build as well.

File: libc/src/__support/fixed_point/fx_bits.h [hunk 2]
```diff
@@ -19,6 +21,36 @@
 
 namespace LIBC_NAMESPACE::fixed_point {
 
+// Bit-wise operations are not available for fixed point types yet.
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, T>
+bit_and(T x, T y) {
+  using BitType = typename FXRep<T>::StorageType;
+  static_assert(sizeof(BitType) * CHAR_BIT == sizeof(T) * CHAR_BIT);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:29]: The two `* CHAR_BIT` are unnecessary.

File: libc/src/__support/fixed_point/fx_bits.h [hunk 3]
```diff
@@ -19,6 +21,36 @@
...
+  BitType x_bit = cpp::bit_cast<BitType>(x);
+  BitType y_bit = cpp::bit_cast<BitType>(y);
+  return cpp::bit_cast<T, BitType>(x_bit & y_bit);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:32]: `cpp::bit_cast<T, BitType>` -> `cpp::bit_cast<T>`

File: libc/src/__support/fixed_point/fx_bits.h [hunk 4]
```diff
@@ -19,6 +21,36 @@
...
+}
+
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, T>
+bit_or(T x, T y) {
+  using BitType = typename FXRep<T>::StorageType;
+  static_assert(sizeof(BitType) * CHAR_BIT == FXRep<T>::TOTAL_LEN);
+  BitType x_bit = cpp::bit_cast<BitType>(x);
+  BitType y_bit = cpp::bit_cast<BitType>(y);
+  return cpp::bit_cast<T, BitType>(x_bit | y_bit);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:42]: same here

File: libc/src/__support/fixed_point/fx_bits.h [hunk 5]
```diff
@@ -19,6 +21,36 @@
...
+}
+
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, T>
+bit_not(T x) {
+  using BitType = typename FXRep<T>::StorageType;
+  static_assert(sizeof(BitType) * CHAR_BIT == sizeof(T) * CHAR_BIT);
+  BitType x_bit = cpp::bit_cast<BitType>(x);
+  return cpp::bit_cast<T, BitType>(~x_bit);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:51]: same here

File: libc/src/__support/fixed_point/fx_bits.h [hunk 6]
```diff
@@ -19,6 +21,36 @@
...
+  BitType x_bit = cpp::bit_cast<BitType>(x);
+  BitType y_bit = cpp::bit_cast<BitType>(y);
+  return cpp::bit_cast<T, BitType>(x_bit | y_bit);
+}
+
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, T>
+bit_not(T x) {
+  using BitType = typename FXRep<T>::StorageType;
+  static_assert(sizeof(BitType) * CHAR_BIT == sizeof(T) * CHAR_BIT);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:49]: same here.

File: libc/src/__support/fixed_point/fx_bits.h [hunk 7]
```diff
@@ -19,6 +21,36 @@
...
+  BitType x_bit = cpp::bit_cast<BitType>(x);
+  BitType y_bit = cpp::bit_cast<BitType>(y);
+  return cpp::bit_cast<T, BitType>(x_bit & y_bit);
+}
+
+template <typename T>
+LIBC_INLINE constexpr cpp::enable_if_t<cpp::is_fixed_point_v<T>, T>
+bit_or(T x, T y) {
+  using BitType = typename FXRep<T>::StorageType;
+  static_assert(sizeof(BitType) * CHAR_BIT == FXRep<T>::TOTAL_LEN);
```
Comment (gchatelet) [inline libc/src/__support/fixed_point/fx_bits.h:39]: Why something different here ?

Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:39]: Oops, those `static_assert` are for debugging that I forgot to remove.  The issue was that for some reason, `bit_cast<fixed point>(x_bits)` cannot infer `BitType`, and keep putting `int` there and failed.  That's why I put the `From` explicitly to resolve the issue.

**Review by nickdesaulniers — APPROVED** [review]

File: libc/src/__support/fixed_point/CMakeLists.txt (hunk 1 above)
Comment (lntue) [reply libc/src/__support/fixed_point/CMakeLists.txt:20]: Done.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 2 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h]: Removed debugging assertion.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 3 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:32]: Added comments explaining the reason.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 4 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:42]: Added comments explaining the reason.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 6 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h]: Removed debugging assertion.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 5 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h:51]: Added comments explaining the reason.

File: libc/src/__support/fixed_point/fx_bits.h (hunk 7 above)
Comment (lntue) [reply libc/src/__support/fixed_point/fx_bits.h]: Removed debugging assertion.

**Pull request merged**